experiment/contractNipopow.sol

```diff
@@ -398,31 +254,99 @@ contract Crosschain {
+                    return false;
+                }
+            }
+        }
+        return true;
+    }
+
     function submitContestingProof(
-        bytes32[4][] memory headers,
-        bytes32[] memory siblings,
-        bytes32[4] memory blockOfInterest
+        bytes32[4][] memory existingHeaders,
+        uint256 lca,
+        bytes32[4][] memory contestingHeaders,
+        bytes32[] memory contestingSiblings,
+        uint256 blockOfInterestIndex
     ) public returns (bool) {
-        bytes32 hashedBlock = hashHeader(blockOfInterest);
-
-        if (events[hashedBlock].expire <= block.number) {
-            return false;
+        require(
+            existingHeaders.length > blockOfInterestIndex &&
+                blockOfInterestIndex >= 0,
+            "Block of interest index is out of range"
+        );
+
+        bytes32 blockOfInterestHash = hashHeader(
+            existingHeaders[blockOfInterestIndex]
+        );
+
+        require(
+            events[blockOfInterestHash].expire > block.number,
+            "Contesting period has expired"
+        );
+
+        require(existingHeaders.length > lca, "Lca out of range");
+
+        require(
+            lca > blockOfInterestIndex,
+            "Block of interest exists in sub-chain"
+        );
+
+        require(
+            events[blockOfInterestHash].proofHash == hashProof(existingHeaders),
+            "Wrong existing proof"
+        );
+
+        // get contesting hashed headers
+        bytes32[] memory contestingHeadersHashed = new bytes32[](
+            contestingHeaders.length
+        );
+        for (uint256 i = 0; i < contestingHeaders.length; i++) {
+            contestingHeadersHashed[i] = hashHeader(contestingHeaders[i]);
         }
-
-        if (
-            !verify(
-                events[hashedBlock].proof,
-                headers,
-                siblings,
-                blockOfInterest
-            )
-        ) {
-            events[hashedBlock].expire = 0;
-            msg.sender.transfer(z);
-            return true;
+        validateInterlink(
+            contestingHeaders,
+            contestingHeadersHashed,
+            contestingSiblings
+        );
+
+        // get existing hashed headers
+        bytes32[] memory existingHeadersHashed = new bytes32[](
+            existingHeaders.length
+        );
+        for (uint256 i = 0; i < existingHeaders.length; i++) {
+            existingHeadersHashed[i] = hashHeader(existingHeaders[i]);
         }
-
-        return false;
+        require(
+            existingHeadersHashed[lca] ==
+                contestingHeadersHashed[contestingHeadersHashed.length - 1],
+            "Wrong lca"
+        );
+
+        require(
+            allDifferent(existingHeadersHashed, contestingHeadersHashed, lca),
+            "Contesting proof[1:] is not different from existing[lca+1:]"
+        );
+
+        // We can ask the caller to provide the level for their proof
```

Comment:
but we don't?

Reply:
Not sure if this is to out advantage. Even if the caller provides the level, we still need to linearly scan the whole proof to check the level. I think this is unworthy because it complicates the interface and only saves a small amount of gas.

Reply:
then rm the comment